src/python/pants/option/config.py

```diff
@@ -230,10 +244,173 @@ def get_value(self, section: str, option: str) -> Optional[str]:
     return self.parser.defaults()
 
 
+_TomlPrimitve = Union[bool, int, float, str]
+_TomlValue = Union[_TomlPrimitve, List[_TomlPrimitve]]
+
+
+@dataclass(frozen=True)
+class _TomlValues(_ConfigValues):
+  values: Dict[str, Any]
+
+  @staticmethod
+  def _section_explicitly_defined(section_values: Dict) -> bool:
+    """Because of the way TOML stores the config as a nested dictionary, it naively appears that
+    certain sections are defined when really only a subscope is defined.
+
+    For example, the user may have specified `cache.java` but not `cache`.
+    """
+    at_least_one_option_defined = any(
+      not isinstance(section_value, dict) for section_value in section_values.values()
+    )
+    blank_section = len(section_values.values()) == 0
+    return at_least_one_option_defined or blank_section
+
+  def _find_section(self, section: str) -> Optional[Dict]:
+    def recurse(mapping: Dict, *, remaining_sections: List[str]) -> Optional[Dict]:
+      if not remaining_sections:
+        return None
+      current_section = remaining_sections[0]
+      if current_section not in mapping:
+        return None
+      section_values = mapping[current_section]
+      if len(remaining_sections) > 1:
+        return recurse(section_values, remaining_sections=remaining_sections[1:])
+      if not self._section_explicitly_defined(section_values):
+        return None
+      return cast(Dict, section_values)
+
+    return recurse(mapping=self.values, remaining_sections=section.split("."))
+
+  def _possibly_interpolate_value(self, raw_value: str) -> str:
+    """For any values with %(foo)s, substitute it with the corresponding default val."""
+    def format_str(value: str) -> str:
+      # Because dictionaries use the symbols `{}`, we must proactively escape the symbols so that
+      # .format() does not try to improperly interpolate.
+      escaped_str = value.replace("{", "{{").replace("}", "}}")
+      new_style_format_str = re.sub(
+        pattern=r"%\((?P<interpolated>[a-zA-Z_0-9]*)\)s",
+        repl=r"{\g<interpolated>}",
+        string=escaped_str,
+      )
+      return new_style_format_str.format(**self.defaults)
+
+    def recursively_format_str(value: str) -> str:
+      # It's possible to interpolate with a value that itself has an interpolation. We must fully
+      # evaluate all expressions for parity with configparser.
+      formatted_str = format_str(value)
+      if not re.search(r"%\([a-zA-Z_0-9]*\)s", formatted_str):
+        return formatted_str
+      return recursively_format_str(formatted_str)
+
+    return recursively_format_str(raw_value)
+
+  def _stringify_val(
+    self, raw_value: _TomlValue, *, interpolate: bool = True, list_prefix: Optional[str] = None,
+  ) -> str:
+    """For parity with configparser, we convert all values back to strings, which allows us to
+    avoid upstream changes to files like parser.py.
+
+    This is clunky. If we drop INI support, we should remove this and use native values."""
```

Comment:
Originally, I did not take this approach and preserved the parsed data types. But, I realized for the initial prototype that this complicates things too much because it means we have a leaky API. For example, it would be really hard to preserve the `my_list_option.append` and `my_list_option.filter` information. 

If we end up removing INI, we could remove this all.